### PR TITLE
featurization refactor

### PR DESCRIPTION
**Pull request by SenWu**

### CODE REVIEW

File: src/fonduer/features/feature_libs/feature_extractors.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    "structural": extract_structural_features,
+    "tabular": extract_tabular_features,
+    "visual": extract_visual_features,
+}
+
+
+class FeatureExtractor(object):
+    """A class to extract features from candidates.
+
+    :param features: a list of feature types which provides in Fonduer, defaults
```
Comment (lukehsiao) [inline src/fonduer/features/feature_libs/feature_extractors.py]:
```suggestion
    :param features: a list of which Fonduer feature types to extract, defaults
```

File: src/fonduer/features/feature_libs/feature_extractors.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+            if feature not in FEATURES:
+                raise ValueError(f"Unrecognized feature type: {feature}")
+            self.feature_extractors.append(FEATURES[feature])
+
+        self.feature_extractors.extend(customize_feature_funcs)
+
+    def extract(self, candidates):
+        """Extract features from candidates.
+
+        :param candidates: A list of candidates that needs to extract features
```
Comment (lukehsiao) [inline src/fonduer/features/feature_libs/feature_extractors.py]:
```suggestion
        :param candidates: A list of candidates to extract features from
```

File: src/fonduer/features/feature_libs/structural_features.py [hunk 3]
```diff
@@ -19,7 +19,12 @@
 binary_strlib_feats = {}
 
 
-def get_structural_feats(candidates):
+def extract_structural_features(candidates):
+    """Extract structural features.
+
+    :param candidates: A list of candidates that needs to extract features
```
Comment (lukehsiao) [inline src/fonduer/features/feature_libs/structural_features.py]:
```suggestion
    :param candidates: A list of candidates to extract features from
```

File: src/fonduer/features/feature_libs/tabular_features.py [hunk 4]
```diff
@@ -17,7 +17,12 @@
 settings = get_config()
 
 
-def get_table_feats(candidates):
+def extract_tabular_features(candidates):
+    """Extract tabular features.
+
+    :param candidates: A list of candidates that needs to extract features
```
Comment (lukehsiao) [inline src/fonduer/features/feature_libs/tabular_features.py]:
```suggestion
    :param candidates: A list of candidates to extract features from
```

File: src/fonduer/features/feature_libs/textual_features.py [hunk 5]
```diff
@@ -29,7 +30,12 @@
 settings = get_config()
 
 
-def get_content_feats(candidates):
+def extract_textual_features(candidates):
+    """Extract textual features.
+
+    :param candidates: A list of candidates that needs to extract features
```
Comment (lukehsiao) [inline src/fonduer/features/feature_libs/textual_features.py]:
```suggestion
    :param candidates: A list of candidates to extract features from
```

File: src/fonduer/features/feature_libs/visual_features.py [hunk 6]
```diff
@@ -16,7 +16,12 @@
 binary_vizlib_feats = {}
 
 
-def get_visual_feats(candidates):
+def extract_visual_features(candidates):
+    """Extract visual features.
+
+    :param candidates: A list of candidates that needs to extract features
```
Comment (lukehsiao) [inline src/fonduer/features/feature_libs/visual_features.py]:
```suggestion
    :param candidates: A list of candidates to extract features from
```

File: CHANGELOG.rst [hunk 7]
```diff
@@ -1,6 +1,37 @@
 [Unreleased]
 ------------
 
+Changed
```
Comment (lukehsiao) [inline CHANGELOG.rst]:
```suggestion
Added
```

File: CHANGELOG.rst [hunk 8]
```diff
@@ -1,6 +1,37 @@
...
+^^^^^^^
```
Comment (lukehsiao) [inline CHANGELOG.rst]:
```suggestion
^^^^^
```

**Comment by HiromuHota** [comment]
Great new feature! I was wondering how I could customize `Featurizer`.
I'd just like "the paper" mentioned in the CHANGELOG to be more specific for those who do not know what this paper is.

**Pull request merged**